[Reporting] Document Network Policy configuration (#80431)
elastic/kibana: docs/user/reporting/network-policy.asciidoc

[role="xpack"]
[[reporting-network-policy]]
=== Restrict requests with a Reporting network policy

> When {report-features} generates PDF reports, it uses the Chromium browser to fully load the {kib} page on the server. This

Reviewer: This still doesn't sound right.  Does this work?

```suggestion
When Reporting generates PDF reports, it uses the Chromium browser to fully load the {kib} page on the server. This
```

[role="xpack"]
[[reporting-network-policy]]
=== Restricting requests with a Reporting Network Policy

> When {report-features} generates PDF reports, it uses the Chromium browser to fully load the Kibana page on the server. This

Reviewer 2: ```suggestion
When {report-features} generates PDF reports, it uses the Chromium browser to fully load the {kib} page on the server. This
```

potentially involves sending requests to external hosts, for example, a request may go to an external image server for showing a
field formatted as an image, or to show an image in a Markdown visualization.

If the Chromium browser is requested to send a request that violates the network policy, Reporting will stop processing the page
before the request goes out, and the report will be marked as a failure. Additional information about the event can be found in

> Kibana's server logs.

Reviewer: ```suggestion
the Kibana server logs.
```

A network policy applies not only to outgoing requests, but also incoming responses. That means if a request goes out to an allowed

> host, but is redirected and a response returns from a denied host, the response will be denied, and the report will fail.

Reviewer 2: nit: Isn't this still technically an outgoing request? The redirect would involve a response from the allowed host, and then the browser would issue a new request to the denied host -- the network policy would then prevent the connection to the denied host.

The way it's currently written, I could misinterpret this to believe that Chromium would connect to the denied host, but simply ignore the response.

Author: Yes, I mistook what is going on in the response handler code that I have become slightly familiar with.

I should have realized the point you made, but now I'm less sure about what the code is doing  :)

Reviewer 3: Yup, that's correct: we do a best effort prior to the request being sent, but if a DNS hijack or something else occurs, this can redirect to internal-only IPs or worse (the `file://` protocol, etc).

This is why we check both the request itself, as well as the response, and if any violate network policies then the session is shutdown immediately (the browser is closed, we log the reason, and the reporting job is failed).

Author: Cool, glad to understand it better. 

I think that level of detail might not be required for the docs. I'm just going to wipe out these sentences.

[NOTE]
============

> Kibana installations are not designed to be publicly accessible over the Internet. The Reporting network policy and other capabilities

Reviewer 2: ```suggestion
{kib} installations are not designed to be publicly accessible over the Internet. The Reporting network policy and other capabilities
```

Is this true? ESS is publicly accessible, you just need our security features enabled to protect the installation

of the Elastic Stack security features do not change this condition.
============

==== Configuring Reporting Network Policy

You configure the network policy by specifying the `xpack.reporting.capture.networkPolicy.rules` setting in `kibana.yml`. A policy is specified as
an array of objects that describe what to allow or deny based on an optionally-provided host and/or protocol. If a host or protocol
is not specified, the rule will match any host or protocol, respectively.

The rule objects are evaluated sequentially from the beginning to the end of the array, and continue until there is a matching rule.
If none of the rules allow a request, the request will be denied.

[source,yaml]
-------------------------------------------------------
# Only allow requests to placeholder.com
xpack.reporting.capture.networkPolicy:
  rules: [ { allow: true, host: "placeholder.com" } ] 
-------------------------------------------------------

[source,yaml]
-------------------------------------------------------
# Only allow requests to https://placeholder.com 
xpack.reporting.capture.networkPolicy:

>   rules: [ { allow: true, host: "placeholder.com", protocol: "https:" } ] 

Author: @joelgriffith do you happen to know if the protocol string must always end with `:` ? I see it that way in the unit tests, but haven't delved deeper.

If that is the case, we should add schema validation to ensure it gets entered right.

Reviewer 3: Yup, it should, likely an implementation bleed in how we parse the request. Schema validation would be great to add, IIRC this was prior to np, so I wonder if there's a better way to do so now

Author: https://github.com/elastic/kibana/pull/80766

-------------------------------------------------------

A final `allow` rule with no host or protocol will allow all requests that are not explicitly denied.

[source,yaml]
-------------------------------------------------------
# Denies requests from http://placeholder.com, but anything else is allowed.
xpack.reporting.capture.networkPolicy:
  rules: [{ allow: false, host: "placeholder.com", protocol: "http:" }, { allow: true }];
-------------------------------------------------------

A network policy can be composed of multiple rules.

[source,yaml]
-------------------------------------------------------
# Allow any request to http://placeholder.com but for any other host, https is required
xpack.reporting.capture.networkPolicy
  rules: [
    { allow: true, host: "placeholder.com", protocol: "http:" },
    { allow: true, protocol: "https:" },
  ]
-------------------------------------------------------

[NOTE]
============

> The `file:` protocol will always be denied, even if there is no network policy configured.

Reviewer: ```suggestion
The `file:` protocol is always denied, even if no network policy is configured.
```

============

> ==== Disabling Reporting Network Policy

Reviewer: ```suggestion
==== Disable a Reporting network policy
```

You can use the `xpack.reporting.capture.networkPolicy.enabled: false` setting to disable the network policy feature. The default for

> this configuration property is `true`, so it is not necessary to explicitly enable it.  

Reviewer 2: **question** is there a functional difference between disabling the network policy, and not defining a network policy at all?
